### DISCUSSION TITLE
Create ICommand Failed

### DISCUSSION THREAD

**Discussion by FlytotheMuum** [unoplatform/uno#11811] [Q&A]
### Current behavior

Templates Used: https://www.nuget.org/packages/Uno.Templates 4.8.8#

### ModelView
    public List<Device> DeviceCollection = new() {
        new Device 
        { 
            Title="123",
            BtnCommand = new SimpleCommand(()=> { 
            //run some command
            })
        }
    };
    public class Device
    {
        public string? Title { get; set; }
        public ICommand? BtnCommand { get; set; } 
    }
    /// <summary>
    /// Simple naive implementation of ICommand for demo purposes
    /// </summary>
    internal class SimpleCommand : ICommand
    {
        private readonly Action _action;

        public SimpleCommand(Action action)
        {
            _action = action;
        }

        public bool CanExecute(object? parameter)
        {
            return true;
        }

        public void Execute(object? parameter)
        {
            _action();
        }
        public event EventHandler? CanExecuteChanged;
    }

    /// <summary>
    /// Simple naive implementation of ICommand for demo purposes
    /// </summary>
    internal class SimpleCommand<T> : ICommand where T : class
    {
        private readonly Action<T> _action;

        public SimpleCommand(Action<T> action)
        {
            _action = action;
        }

        public bool CanExecute(object? parameter)
        {
            return parameter is T;
        }

        public void Execute(object? parameter)
        {
            if (parameter is T value)
            {
                _action(value);
            }
        }
        public event EventHandler? CanExecuteChanged;
    }

### Xaml
                    <Grid>
                        <GridView 
                            SelectionMode="None" IsItemClickEnabled="True"
                            ItemsSource="{Binding DeviceCollection }">
                            <GridView.ItemTemplate>
                                <DataTemplate>
                                    <Button 
                                    Content="{Binding Title}" 
                                    Command="{Binding BtnCommand }" 
                                    CommandParameter="{Binding Title}" />
                                </DataTemplate>
                            </GridView.ItemTemplate>
                        </GridView>
                    </Grid>

Program can't build with fail CS0067

### Expected behavior

Build Sucessful

### How to reproduce it (as minimally and precisely as possible)

[.NET CLI](https://www.nuget.org/packages/Uno.Templates#dotnet-cli)
```
dotnet new install Uno.Templates::4.8.8
dotnet new unoapp
```

### Workaround

......
#pragma warning disable CS0067
public event EventHandler? CanExecuteChanged;
......

### Works on UWP/WinUI

None

### IDE

Visual Studio 2022

### IDE version

17.6 Preview 2.0

**Comment by jeromelaban** [comment]
Thanks for the report. 

Let's discuss first here, as it seems that the behavior you're observing is appropriate, since the member really is not used. Note that the default template [uses `TreatWarningsAsErrors` by default](https://github.com/unoplatform/uno.extensions/blob/8d491ea360dda7721afc9bf872dc5fe6db35bc68/src/Uno.Extensions.Templates/content/unoapp-extensions/Directory.Build.props#L15).

**Comment by Youssef1313** [comment]
I guess `SimpleCommand` isn't included in the template and it's code you wrote yourself? In that case, the behavior is correct. This warning is coming from the compiler *and* the template treats warnings as errors, as @jeromelaban mentioned.

Possible ways to avoid the issue (ordered in *my* preference):

- Use `public event EventHandler? CanExecuteChanged { add { } remove { } }`
- `#pragma warning disable CS0067`
- Write a custom `DiagnosticSuppressor` that disables this diagnostic interface-implemented events. This could be done in Uno itself, but since this is the less preferred option for me, I don't think we should do it. Though I definitely would defer to @jeromelaban on that in case he thinks otherwise.

**Reply by FlytotheMuum** [reply]
Hi there, Im sure This is the best way to fix this issue:
`Use public event EventHandler? CanExecuteChanged { add { } remove { } }`

I Get SimpleCommand  from sample code:
[uno-bootcamp/modules/03-Let-views-do-views/TodoApp/TodoApp.Shared/ViewModels/SimpleCommand.cs](https://github.com/unoplatform/workshops/blob/98aa8e6026e77028ecae447f3fc8edd4df4b5a53/uno-bootcamp/modules/03-Let-views-do-views/TodoApp/TodoApp.Shared/ViewModels/SimpleCommand.cs)

The reason to use SimpleCommand is that I want the buttons in the Card to call the ICommand method sucessfully.

**Reply by FlytotheMuum** [reply]
```
<utu:Card HeaderContent="Outlined card"
          SupportingContent="{Binding DeviceCollection}"
          AutomationProperties.AutomationId="Outlined_Card_Supplemental_Action_Button">
   <utu:Card.SupportingContentTemplate>
      <DataTemplate>


               <Button 
                       Command="{Binding BtnCommand}"
                       AutomationProperties.AutomationId="Action_1_Button" 
                />


      </DataTemplate>
   </utu:Card.SupportingContentTemplate>
</utu:Card>
```
